transaction/std/src/memo_builder/gift_code_sender_memo_builder.rs

```diff
@@ -0,0 +1,276 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Defines the Memo Builder for the gift code sender memo (0x0002)
+//! specified in MCIP #32
+
+use crate::{
+    memo::{GiftCodeSenderMemo, UnusedMemo},
+    MemoBuilder, ReservedDestination,
+};
+use mc_account_keys::PublicAddress;
+use mc_transaction_core::{Amount, MemoContext, MemoPayload, NewMemoError};
+
+/// There are three possible gift code memo types specified in MCIP #32
+/// | Memo type bytes | Name                        |
+/// | -----------     | -----------                 |
+/// | -->0x0002<--    | Gift Code Sender Memo       |
+/// |    0x0201       | Gift Code Funding Memo      |
+/// |    0x0202       | Gift Code Cancellation Memo |
+/// This memo builder builds a gift code sender memo (0x0002). A gift code
+/// considered redeemed when the Receiver uses the TxOut spend private key
+/// of the gift code TxOut they received from the Sender to send the TxOut
+/// from the sender's gift code subaddress to their own change subaddress.
+/// The destination memo is written into that TxOut at the change address
+/// and includes an optional Utf-8 note up to 64 bytes long the Receiver
+/// can use to record any information they desire about the gift code.
+#[derive(Clone, Debug)]
+pub struct GiftCodeSenderMemoBuilder {
+    // Utf-8 formatted note within the memo that can be up to 64 bytes long
+    note: String,
+    // Whether or not to enable change memos
+    gift_code_change_memo_enabled: bool,
+    // Whether we've already written the change memo
+    wrote_change_memo: bool,
+    // Whether our last note setting attempt was invalid and the note hasn't
+    // been reset with a valid note or cleared
+    attempted_invalid_note: bool,
```

Comment:
I don't see what this gives us. Its the callers responsibility to handle a failure of `set_gift_code_sender_note` and not continue building the transaction.

Reply:
I'm not sure how the extra redundancy hurts either. Given the records on chain are immutable, another check doesn't hurt.